File: preview/classics/preview.ml
(* ======================================
   OCaml Language Semantic Preview
   OCaml 5+
   ====================================== *)


(* ---------- Variables ---------- *)

let nome : string = "Esron"

let idade : int = 30


(* Inferência de tipos *)

let linguagem = "OCaml"



(* ---------- Basic Types ---------- *)

let numero : int = 100

let decimal : float = 3.14

let ativo : bool = true



(* ---------- Tuples ---------- *)

let pessoa =
  ("Esron", 30)



(* ---------- Lists ---------- *)

let linguagens =
  [
    "OCaml";
    "F#";
    "Haskell"
  ]



(* ---------- Functions ---------- *)

let saudacao nome =
  "Olá " ^ nome



let soma a b =
  a + b



(* ---------- Recursive Function ---------- *)

let rec fatorial n =

  if n <= 1 then

    1

  else

    n * fatorial (n - 1)



(* ---------- Pattern Matching ---------- *)

let verificar valor =

  match valor with

  | 0 ->
      "zero"

  | 1 ->
      "um"

  | _ ->
      "outro"



(* ---------- Variant Types ---------- *)

type status =

  | Ativo
  | Inativo
  | Bloqueado



let mensagem estado =

  match estado with

  | Ativo ->
      "Usuário ativo"

  | Inativo ->
      "Usuário desligado"

  | Bloqueado ->
      "Usuário bloqueado"



(* ---------- Records ---------- *)

type usuario = {

  nome : string;

  idade : int;

}



let criar_usuario nome idade =

  {

    nome = nome;

    idade = idade;

  }



(* ---------- Modules ---------- *)

module Calculadora = struct


  let multiplicar a b =

    a * b


end



(* ---------- Functors Style ---------- *)

module type Numero = sig

  val valor : int

end



(* ---------- Higher Order Function ---------- *)

let aplicar func valor =

  func valor



let dobro x =

  x * 2



(* ---------- Pipeline Style ---------- *)

let resultado =

  10

  |> dobro



(* ---------- Exceptions ---------- *)

exception ErroPersonalizado of string



let validar idade =

  if idade < 0 then

    raise
      (ErroPersonalizado "Idade inválida")

  else

    idade



(* ---------- Main ---------- *)

let () =

  print_endline
    (saudacao nome);

  print_int
    (soma 10 20)

File: preview/classics/preview.ex
# ======================================
# Elixir Language Semantic Preview
# Elixir 1.17+
# ======================================


# ---------- Variables ----------
# Elixir usa imutabilidade

nome = "Esron"

idade = 30


# ---------- Basic Types ----------

numero = 100

decimal = 3.14

ativo = true

lista = [
  "Elixir",
  "Phoenix",
  "Erlang"
]

mapa = %{
  nome: "Esron",
  idade: 30
}



# ---------- Functions ----------

defmodule Saudacao do

  def ola(nome) do

    "Olá, #{nome}"

  end


  def soma(a, b) do

    a + b

  end

end



# ---------- Module ----------

defmodule Pessoa do


  defstruct [
    :nome,
    :idade
  ]


  def apresentar(
        %Pessoa{
          nome: nome,
          idade: idade
        }
      ) do

    "#{nome} tem #{idade} anos"

  end


end



# ---------- Pattern Matching ----------

defmodule Usuario do


  def status(
        %{ativo: true}
      ) do

    "Usuário ativo"

  end


  def status(
        %{ativo: false}
      ) do

    "Usuário inativo"

  end


end



# ---------- Case ----------

resultado =
  case idade do

    30 ->
      "Trinta anos"


    _ ->
      "Outra idade"

  end



# ---------- Anonymous Function ----------

multiplicar =
  fn a, b ->

    a * b

  end



# ---------- Pipe Operator ----------

texto =
  "elixir"


texto
|> String.upcase()
|> IO.puts()



# ---------- Enum Functional Programming ----------

numeros =
  [1, 2, 3, 4, 5]


dobrados =
  Enum.map(
    numeros,
    fn numero ->

      numero * 2

    end
  )



# ---------- Recursion ----------

defmodule Contador do


  def contar(0) do

    :fim

  end


  def contar(numero) do

    IO.puts(numero)

    contar(numero - 1)

  end


end



# ---------- Protocol ----------

defprotocol Mostrar do

  def exibir(valor)

end



# ---------- Struct Usage ----------

pessoa =
  %Pessoa{
    nome: "Esron",
    idade: 30
  }



IO.puts(
  Pessoa.apresentar(pessoa)
)



# ---------- Concurrency ----------

spawn(fn ->

  IO.puts(
    "Processo Elixir"
  )

end)



# ---------- Main Execution ----------

IO.puts(
  Saudacao.ola(nome)
)

IO.puts(
  Saudacao.soma(10,20)
)